### PR TITLE
#15165: Pre-release QA

### PR DESCRIPTION
**Pull request by jnovinger**
### Closes: #15165

Pre-release QA found the 802.1Q Mode field on `InterfaceForm` and `VMInterfaceForm` no longer triggers a partial swap: changing mode fires no HTMX request, so dependent VLAN fields never update. This is the exact case #15165 delivered; it regressed on `feature`. The root cause turned out to be a general trap, so the PR also guards the whole class.

- Cause: #21712 declared `mode` as an explicit field, which shadows `Meta.widgets` (Django discards `Meta.widgets`/`labels`/`help_texts` for explicitly declared fields), so the `HTMXSelect` set there was dropped and `mode` rendered as a plain `Select`.
- Set `HTMXSelect` on the explicit field; removed the dead `Meta` entries. `HTMXSelect` now includes `AttrSelectMixin` so `mode` keeps both the partial swap and its option descriptions, and `AttrSelectMixin.__deepcopy__` stops per-instance widget copies sharing one `descriptions` dict.
- Fixed the same shadow on `VirtualChassisForm.master` (`SelectWithPK` was dropped, rendering a plain `Select` with no PK in the label) and removed three dead shadowed `Meta` entries elsewhere.
- `HTMXPartialSwapRenderingTestCase` pins every partial-swap field plus the intentional full-form fields, asserting on the widget bound to each field. `MetaShadowingTestCase` walks every `ModelForm` and fails on any `Meta` key that collides with a declared field, so this class can't recur unnoticed. `ConfigRevisionForm` is allowlisted (its metaclass sources widgets from `Meta` by design); its monospace-class drop is a follow-up.

### CODE REVIEW

**Comment by jnovinger** [comment]
@claude /review Only report on issues of potential concern; we don't need commentary on anything that doesn't require further consideration.

Context for the review: the commit message explains the widget-shadowing root cause. Two areas worth scrutiny: (1) adding `AttrSelectMixin` to `HTMXSelect`'s bases (MRO, and whether any existing `HTMXSelect` field without descriptions is affected), and (2) whether `HTMXPartialSwapRenderingTestCase` correctly resolves the bound widget for both the plain `HTMXSelect` path and the `GenericObjectSelect` MultiWidget path.

**Comment by jnovinger** [comment]
@claude /review Only report on issues of potential concern; skip anything that doesn't need further consideration.

Addressed in `c948738`:
- `__deepcopy__` on `AttrSelectMixin` so per-instance copies own their `descriptions` dict.
- Removed the dead `mode` entries from `Meta.labels` on both forms.
- Partial-swap tests now assert an `hx-get` verb, not just `hx-target`/`hx-select`.
- `_hx_widget` selects the subwidget by `isinstance(HTMXSelect)` instead of a hardcoded index.
- Moved `HTMXPartialSwapRenderingTestCase` to `netbox/tests/`.

Deferring the generalized `_warn_missing_htmx_fieldsets`: covering the plain-`HTMXSelect` path means moving the check off `GenericObjectFormMixin` onto a base form class, which is broader than this fix. Follow-up.

**Comment by jnovinger** [comment]
@claude /review Only report on issues of potential concern; skip anything that doesn't need further consideration.

Addressed in `8a8813f`:
- Added `RackForm.rack_type` and `FrontPortCreateForm.device` to `FULL_FORM_FIELDS`; both hand-roll the full-form attrs.
- Each partial-swap target is now asserted against the form's declared `FieldSet` html_ids, so a typo'd `hx_target_id` fails the test. `CableForm` is exempt: its targets are template `<div>`s, not `FieldSet`s.
- The MultiWidget subwidget lookup gets a default and a descriptive assert instead of a bare `StopIteration`.
- `HTMXSelect` now derives from the local `Select`.
- Moved the `InterfaceForm` Meta note to Meta body level.

**Comment by jnovinger** [comment]
This started as the #15165 regression fix and grew to cover the bug class behind it. Summary of what's here and what's deliberately not.

Fixed:
- `InterfaceForm.mode` / `VMInterfaceForm.mode` 802.1Q partial swap (the reported regression). `HTMXSelect` gained `AttrSelectMixin` so `mode` keeps both the swap and its option descriptions.
- `VirtualChassisForm.master`: `SelectWithPK` was shadowed the same way and rendered a plain `Select` (no PK in the label). Moved onto the declared field.
- Removed three dead shadowed `Meta` entries (`CustomFieldForm.type` help_text, `EventRuleForm.conditions` widget, `WirelessLinkForm` labels) that had no effect only because the declared field already matched.
- `AttrSelectMixin.__deepcopy__` so per-instance widget copies own their `descriptions` dict.

Tests:
- `HTMXPartialSwapRenderingTestCase` pins the 16 partial-swap fields and the intentional full-form fields.
- `MetaShadowingTestCase` walks every `ModelForm` and fails on any `Meta.widgets/labels/help_texts` key that collides with a declared field, so this class of bug can't recur unnoticed.

Not done, deliberately:
- `ConfigRevisionForm` has the same shadow, but its `ConfigFormMetaclass` sources widgets from `Meta` for synthesized fields, so the banner/validator textareas lose their monospace class. Fixing it means changing the metaclass. Allowlisted in the test, follow-up.
- Bulk-edit `GenericObjectChoiceField` sets `hx-select=#form_fields` (a third swap category) and isn't covered by either test list. Not a regression, follow-up.

File: netbox/dcim/forms/model_forms.py [hunk 1]
```diff
@@ -2186,11 +2185,9 @@ class Meta:
             'speed': NumberWithOptions(
                 options=InterfaceSpeedChoices
             ),
-            'mode': HTMXSelect(hx_target_id='dot1q-switching'),
-        }
-        labels = {
-            'mode': '802.1Q Mode',
         }
+        # Note: `mode` is configured on the explicitly declared field above (widget and label),
+        # not here. Meta.widgets and Meta.labels are both ignored for explicitly declared fields.
```
Comment (pheus) [inline netbox/dcim/forms/model_forms.py]: Please remove the explanation.

File: netbox/extras/forms/model_forms.py [hunk 2]
```diff
@@ -129,10 +129,8 @@ class Meta:
         model = CustomField
         fields = '__all__'
         help_texts = {
-            'type': _(
-                "The type of data stored in this field. For object/multi-object fields, select the related object "
-                "type below."
-            ),
+            # Note: `type` is explicitly declared above with the same help_text; a Meta.help_texts
+            # entry for it would be silently discarded, so it must not be set here.
```
Comment (pheus) [inline netbox/extras/forms/model_forms.py]: Same here

File: netbox/extras/forms/model_forms.py [hunk 3]
```diff
@@ -641,7 +639,8 @@ class Meta:
             'action_object_type', 'action_object_id', 'action_data', 'owner', 'comments', 'tags'
         )
         widgets = {
-            'conditions': forms.Textarea(attrs={'class': 'font-monospace'}),
+            # Note: `conditions` is explicitly declared above (a JSONField that already renders a
+            # monospace Textarea); a Meta.widgets entry for it would be silently discarded.
```
Comment (pheus) [inline netbox/extras/forms/model_forms.py]: Same here

File: netbox/netbox/tests/test_forms.py [hunk 4]
```diff
@@ -301,3 +317,162 @@ def test_wwn_field_nullable(self):
...
+    the form declares explicitly, the declared field wins and the Meta entry is silently discarded.
+    A key appearing in both is therefore dead config at best and a dropped widget/label/help_text at
+    worst (the #15165 regression, and the VirtualChassis master SelectWithPK before this PR). This
+    walks every ModelForm and asserts the two never overlap, so the next occurrence fails here.
+    """
+    # ConfigRevisionForm builds its parameter fields via a custom metaclass (ConfigFormMetaclass)
+    # that turns them into declared_fields while still sourcing widgets from Meta.widgets. Its
+    # overlap is a known consequence of that design, not the shadowing bug this guards; excluded
+    # here and left for a separate cleanup.
+    ALLOWED = {'ConfigRevisionForm'}
```
Comment (pheus) [inline netbox/netbox/tests/test_forms.py]: Deferring the `ConfigRevisionForm` cleanup seems reasonable, but I don't think we should allowlist the entire form here. `ConfigFormMetaclass` creates the parameter fields from `param.field_kwargs` and adds them to `attrs`; it does not apply `Meta.widgets`. The six monospace widget entries are therefore affected by the same declared-field shadowing this test is intended to catch.

Could we make the exception specific to those known `(form, Meta attribute, field)` combinations and reference the follow-up issue? That would let us defer the cleanup without also exempting any future overlap added to this form.

File: netbox/netbox/tests/test_forms.py [hunk 5]
```diff
@@ -301,3 +317,162 @@ def test_wwn_field_nullable(self):
...
+        # Import every app's forms package so all ModelForm subclasses are registered before walking.
+        import importlib
+        import pkgutil
+        for app in (
+            'circuits', 'core', 'dcim', 'extras', 'ipam', 'tenancy', 'users', 'utilities',
+            'virtualization', 'vpn', 'wireless', 'netbox',
+        ):
+            try:
+                pkg = importlib.import_module(f'{app}.forms')
+            except ImportError:
```
Comment (pheus) [inline netbox/netbox/tests/test_forms.py]: Could we avoid swallowing `ImportError` here and again at line 451? This also catches an import failure raised while executing the package or submodule, so the invariant could pass after silently omitting part of the forms hierarchy.

These are core NetBox modules, so I think an import failure should fail the test. If a missing module is genuinely expected, we could limit the exception to `ModuleNotFoundError` for the exact module being probed.

File: netbox/netbox/tests/test_forms.py [hunk 6]
```diff
@@ -301,3 +317,162 @@ def test_wwn_field_nullable(self):
...
+
+    def test_meta_config_does_not_shadow_declared_fields(self):
+        for form_class in self._all_model_forms():
+            if form_class.__name__ in self.ALLOWED:
+                continue
+            meta = getattr(form_class, 'Meta', None)
+            declared = set(getattr(form_class, 'declared_fields', {}))
+            if meta is None or not declared:
+                continue
+            for attr in ('widgets', 'labels', 'help_texts'):
```
Comment (pheus) [inline netbox/netbox/tests/test_forms.py]: The test is described as guarding the whole shadowing class, but Django takes the same declared-field path for `Meta.error_messages`, `Meta.field_classes`, and `Meta.localized_fields`.

Could we either include those options as well, handling `localized_fields` separately, or narrow the test name, docstring, and PR description to the three attributes covered here?

File: netbox/virtualization/forms/model_forms.py [hunk 7]
```diff
@@ -509,12 +510,8 @@ class Meta:
...
             'owner', 'tags',
         ]
-        labels = {
-            'mode': _('802.1Q Mode'),
-        }
-        widgets = {
-            'mode': HTMXSelect(hx_target_id='dot1q-switching'),
-        }
+        # Note: `mode` is configured on the explicitly declared field above (widget and label),
+        # not here. Meta.widgets and Meta.labels are both ignored for explicitly declared fields.
```
Comment (pheus) [inline netbox/virtualization/forms/model_forms.py]: Same here

File: netbox/wireless/forms/model_forms.py [hunk 8]
```diff
@@ -228,7 +228,5 @@ class Meta:
                 attrs={'data-toggle': 'password'}
             ),
         }
-        labels = {
-            'auth_type': 'Type',
-            'auth_cipher': 'Cipher',
-        }
+        # Note: `auth_type` and `auth_cipher` are explicitly declared with these same labels;
+        # Meta.labels entries for them would be silently discarded, so they are not set here.
```
Comment (pheus) [inline netbox/wireless/forms/model_forms.py]: Same

File: netbox/netbox/tests/test_forms.py [hunk 9]
```diff
@@ -301,3 +317,162 @@ def test_wwn_field_nullable(self):
...
+
+class HTMXPartialSwapRenderingTestCase(TestCase):
+    """
+    Verify that each form field's HTMX swap wiring resolves to the widget actually bound to the
+    field. An explicitly declared form field silently overrides a widget set via Meta.widgets, so
+    a field can have HTMXSelect assigned in Meta yet end up with a plain widget carrying no HTMX
+    wiring. Unit-testing the widget in isolation cannot see that, which is how the interface 802.1Q
+    mode swap regressed unnoticed. These tests read the bound field's widget attrs directly so a
+    shadowed field is caught.
+    """
```
Comment (pheus) [inline netbox/netbox/tests/test_forms.py]: Could we make a broader pass over the comments and docstrings throughout this class? There is useful context here, but the current documentation is considerably more verbose than the surrounding tests. Several blocks explain the regression history or walk through assertions whose intent is already clear from the class, method, and constant names.

I think most of these comments can be removed. Where the full-form and fieldset exceptions genuinely need explanation, a brief sentence describing the lasting invariant should be enough. The detailed history of how the regression occurred is better kept in the PR or commit message.

File: netbox/netbox/tests/test_forms.py [hunk 10]
```diff
@@ -301,3 +317,162 @@ def test_wwn_field_nullable(self):
...
+
+class MetaShadowingTestCase(TestCase):
+    """
+    Guard the whole bug class behind #15165, not just the HTMX fields. Django's ModelFormMetaclass
+    applies Meta.widgets/labels/help_texts only to fields it generates from the model; for a field
+    the form declares explicitly, the declared field wins and the Meta entry is silently discarded.
+    A key appearing in both is therefore dead config at best and a dropped widget/label/help_text at
+    worst (the #15165 regression, and the VirtualChassis master SelectWithPK before this PR). This
+    walks every ModelForm and asserts the two never overlap, so the next occurrence fails here.
+    """
```
Comment (pheus) [inline netbox/netbox/tests/test_forms.py]: The documentation in this class could also be much shorter. A concise description such as “Ensure declared fields do not shadow supported `ModelForm.Meta` configuration” would capture the lasting invariant without embedding the history of #15165 and the earlier `VirtualChassisForm` case in the test.

The `ConfigRevisionForm` explanation should be removed or rewritten as part of addressing that exception, and the import comment can remain a single brief note if the registration side effect is not otherwise apparent.

File: netbox/utilities/forms/widgets/select.py [hunk 11]
```diff
@@ -26,6 +26,13 @@ def __init__(self, *args, descriptions=None, **kwargs):
         super().__init__(*args, **kwargs)
         self.descriptions = descriptions or {}
 
+    def __deepcopy__(self, memo):
+        # Django's ChoiceWidget.__deepcopy__ copies attrs and choices but not descriptions, so
+        # every per-instance widget copy would otherwise share one dict with the class-level widget.
```
Comment (pheus) [inline netbox/utilities/forms/widgets/select.py]: Could we keep the new documentation here focused on behavior rather than the implementation history? The copy operation, the inheritance from `Select`, and the test name already make most of this clear.

I would remove the `__deepcopy__` comment and the test docstring, and reduce the additional `HTMXSelect` documentation to a short description that it supports option descriptions alongside its HTMX behavior.

**Review by pheus — CHANGES_REQUESTED** [review]
Thanks!

I’m requesting changes to make the new invariant reliable: please narrow the `ConfigRevisionForm` exception, avoid hiding import failures, and align the test’s stated scope with what it checks. The added comments and docstrings also need a substantial trim, as many repeat the code or regression history.

Once those points are addressed, I don’t see anything else blocking.

File: netbox/netbox/tests/test_forms.py [hunk 12]
```diff
@@ -301,3 +317,151 @@ def test_wwn_field_nullable(self):
...
+        for form_class in self._all_model_forms():
+            meta = getattr(form_class, 'Meta', None)
+            declared = set(getattr(form_class, 'declared_fields', {}))
+            if meta is None or not declared:
+                continue
+            # All Meta options Django keys by field name and discards for explicitly declared
+            # fields. localized_fields is a list of names rather than a dict, but set() of either a
+            # dict or a list yields the field names, so one comparison covers all six.
+            for attr in ('widgets', 'labels', 'help_texts', 'error_messages', 'field_classes', 'localized_fields'):
+                overlap = declared & set(getattr(meta, attr, None) or {})
```
Comment (pheus) [inline netbox/netbox/tests/test_forms.py]: One edge case remains here: Django permits `Meta.localized_fields = ALL_FIELDS` (`"__all__"`). Converting that sentinel with `set()` yields its characters, so this check would not detect declared model fields.

Could we handle `ALL_FIELDS` explicitly and add a focused regression test?

**Review by pheus — CHANGES_REQUESTED** [review]
Thanks for addressing the previous feedback. The production fix looks good.

I found one remaining edge case in the `localized_fields` check. Once `ALL_FIELDS` is handled, this looks good to me.

**Review by pheus — APPROVED** [review]
Thanks for addressing the remaining `ALL_FIELDS` case. The implementation and regression coverage look good.
